tests/scripts/generate_psa_tests.py

```diff
@@ -305,6 +306,145 @@ def test_cases_for_key_generation(self) -> Iterator[test_case.TestCase]:
+            category: crypto_knowledge.AlgorithmCategory,
+    ) -> Iterator[test_case.TestCase]:
+        """Generate failure test cases for keyless operations with the specified algorithm."""
+        if alg.can_do(category):
+            # Compatible operation, unsupported algorithm
+            for dep in automatic_dependencies(alg.base_expression):
+                yield self.make_test_case(alg, category,
+                                          self.Reason.NOT_SUPPORTED,
+                                          not_deps=frozenset([dep]))
+        else:
+            # Incompatible operation, supported algorithm
+            yield self.make_test_case(alg, category, self.Reason.INVALID)
+
+    def one_key_test_cases(
+            self,
+            alg: crypto_knowledge.Algorithm,
+            category: crypto_knowledge.AlgorithmCategory,
+    ) -> Iterator[test_case.TestCase]:
+        """Generate failure test cases for one-key operations with the specified algorithm."""
+        for kt in self.key_types:
+            key_is_compatible = kt.can_do(alg)
+            if key_is_compatible and alg.can_do(category):
+                # Compatible key and operation, unsupported algorithm
+                for dep in automatic_dependencies(alg.base_expression):
+                    yield self.make_test_case(alg, category,
+                                              self.Reason.NOT_SUPPORTED,
+                                              kt=kt, not_deps=frozenset([dep]))
+                # Public key for a private-key operation
+                if category.is_asymmetric() and kt.is_public():
+                    yield self.make_test_case(alg, category,
+                                              self.Reason.PUBLIC,
+                                              kt=kt)
+            elif key_is_compatible:
+                # Compatible key, incompatible operation, supported algorithm
+                yield self.make_test_case(alg, category,
+                                          self.Reason.INVALID,
+                                          kt=kt)
+            elif alg.can_do(category):
+                # Incompatible key, compatible operation, supported algorithm
+                yield self.make_test_case(alg, category,
+                                          self.Reason.INCOMPATIBLE,
+                                          kt=kt)
+            else:
+                # Incompatible key and operation. Don't test cases where
+                # multiple things are wrong, to keep the number of test
+                # cases reasonable.
+                pass
+
+    def test_cases_for_algorithm(
+            self,
+            alg: crypto_knowledge.Algorithm,
+    ) -> Iterator[test_case.TestCase]:
+        """Generate operation failure test cases for the specified algorithm."""
+        for category in crypto_knowledge.AlgorithmCategory:
+            if category == crypto_knowledge.AlgorithmCategory.PAKE:
```

Comment:
I can't think of a better place to detect the PAKE category. Sure, I could change the code to
```
categories = set(crypto_knowledge.AlgorithmCategory)
categories.discard(crypto_knowledge.AlgorithmCategory.PAKE)
for category in categories:
    …
```
but this doesn't seem any clearer to me. The code to remove PAKE could be made an auxiliary function, but again I don't see an improvement.

`PAKE` does need to be in the enum, since PAKE algorithms exists. You can do `Algorithm("PSA_ALG_JPAKE").category`.

When PAKE is implemented, `test_cases_for_algorithm` will change to something like
```
for category in crypto_knowledge.AlgorithmCategory:
    if category == crypto_knowledge.AlgorithmCategory.PAKE:
        yield from self.pake_test_cases(alg)
    elif category.requires_key():
    …
```
with `pake_test_cases` enumerating primitives and, for compatible primitives, callin `one_key_test_cases`. So the current structure is the long-term structure.